### PR TITLE
Allowing for configurable Database of TestRunner (deduplicating BasicRocksdbTestRunner)

### PR DESCRIPTION
**Pull request by jakrawcz-rdx**
When working on https://github.com/radixdlt/radixdlt-scrypto/pull/1240 I wanted to repeatedly run benchmarks on memoryDb vs rocksDb.
This was problematic, because `BasicRocksdbTestRunner` was very separate from `TestRunner`. But in fact they only differed by their `substate_db` dependency, so I made it configurable at `TestRunner` and then deleted `BasicRocksdbTestRunner`.

### CODE REVIEW

File: scrypto-unit/src/test_runner.rs [hunk 1]
```diff
@@ -235,24 +235,38 @@ impl CustomGenesis {
...
     custom_genesis: Option<CustomGenesis>,
+    custom_extension: E,
+    custom_database: D,
     trace: bool,
     state_hashing: bool,
 }
 
-impl TestRunnerBuilder {
-    pub fn new() -> TestRunnerBuilder {
+impl TestRunnerBuilder<NoExtension, InMemorySubstateDatabase> {
```
Comment (talekhinezh) [inline scrypto-unit/src/test_runner.rs:254]: Cleaner may be `impl DefaultTestRunner {`

File: scrypto-unit/src/test_runner.rs [hunk 2]
```diff
@@ -235,24 +235,38 @@ impl CustomGenesis {
     }
 }
 
-pub struct TestRunnerBuilder {
+pub trait Database:
```
Comment (talekhinezh) [inline scrypto-unit/src/test_runner.rs:238]: Perhaps change name to `TestDatabase` as it's the fact that it's a test that we need all of the listed traits.

**Review by talekhinezh — APPROVED** [review]

File: scrypto-unit/src/test_runner.rs (hunk 1 above)
Comment (jakrawcz-rdx) [reply scrypto-unit/src/test_runner.rs:254]: I had the same thought!
...but this is a `TestRunnerBuilder`, not `TestRunner` :(

File: scrypto-unit/src/test_runner.rs (hunk 2 above)
Comment (jakrawcz-rdx) [reply scrypto-unit/src/test_runner.rs:238]: I always welcome a good naming suggestion 👍

File: scrypto-unit/src/test_runner.rs (hunk 1 above)
Comment (talekhinezh) [reply scrypto-unit/src/test_runner.rs:257]: ah!

**Comment by jakrawcz-rdx** [comment]
@talekhinezh is the `Cancelled after 23m` normal for these ☝️ 2 checks?
(or please click merge if safe)

**Comment by talekhinezh** [comment]
> @talekhinezh is the `Cancelled after 23m` normal for these ☝️ 2 checks? (or please click merge if safe)

Hrm, let me try and rerun. Not normal but sometimes it happens.

**Comment by jakrawcz-rdx** [comment]
I retried it again in the morning and consistently these 2 (`Radix Engine tests (wasmer), Radix Engine tests (k8s-linux-runner)`) are getting cancelled, after 19-24 minutes.
Other current PRs against develop (e.g. https://github.com/radixdlt/radixdlt-scrypto/pull/1256) don't have this problem.

I cannot find any logs (or information on "why cancelled") on CI. Maybe I'm clicking wrong things. How to find out more?

Is `cancelled == timeout`? Can my change even affect the performance?
Or is it about something else?

**Comment by iamyulong** [comment]
Thanks for cleaning this up! 

Wanted to do the same but needed to get something in quick,  so added `BasicRocksdbTestRunner` :-D

**Comment by jakrawcz-rdx** [comment]
Let's see if updating from `develop` helps the CI.

Even if not, I will hit "Merge", backed by this 🟢 :

image

**Comment by jakrawcz-rdx** [comment]
yup, same 2 cancellations, but I am merging anyway, as promised.

**Pull request merged**